codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java

```diff
@@ -0,0 +1,169 @@
+                @Override
+                public CharSequence escape(CharSequence value) {
+                    return StringUtil.escapeCsv(value);
+                }
+            };
+        }
+        return charSequenceEscaper;
+    }
+
+    @Override
+    public CombinedHttpHeaders add(CharSequence name, CharSequence value) {
+        return addEscapedValue(name, StringUtil.escapeCsv(value));
+    }
+
+    @Override
+    public CombinedHttpHeaders add(CharSequence name, CharSequence... values) {
+        return addEscapedValue(name, commaSeparate(charSequenceEscaper(), values));
+    }
+
+    @Override
+    public CombinedHttpHeaders add(CharSequence name, Iterable<? extends CharSequence> values) {
+        return addEscapedValue(name, commaSeparate(charSequenceEscaper(), values));
+    }
+
+    @Override
+    public CombinedHttpHeaders addObject(CharSequence name, Iterable<?> values) {
+        return addEscapedValue(name, commaSeparate(objectEscaper(), values));
+    }
+
+    @Override
+    public CombinedHttpHeaders addObject(CharSequence name, Object... values) {
+        return addEscapedValue(name, commaSeparate(objectEscaper(), values));
+    }
+
+    @Override
+    public CombinedHttpHeaders set(CharSequence name, CharSequence... values) {
+        super.set(name, commaSeparate(charSequenceEscaper(), values));
+        return this;
+    }
+
+    @Override
+    public CombinedHttpHeaders set(CharSequence name, Iterable<? extends CharSequence> values) {
+        super.set(name, commaSeparate(charSequenceEscaper(), values));
+        return this;
+    }
+
+    @Override
+    public CombinedHttpHeaders setObject(CharSequence name, Object... values) {
+        super.set(name, commaSeparate(objectEscaper(), values));
+        return this;
+    }
+
+    @Override
+    public CombinedHttpHeaders setObject(CharSequence name, Iterable<?> values) {
+        super.set(name, commaSeparate(objectEscaper(), values));
+        return this;
+    }
+
+    private CombinedHttpHeaders addEscapedValue(CharSequence name, CharSequence escapedValue) {
+        CharSequence currentValue = super.get(name);
+        if (currentValue == null) {
+            super.add(name, escapedValue);
+        } else {
+            super.set(name, commaSeparateEscapedValues(currentValue, escapedValue));
+        }
+        return this;
+    }
+
+    private <T> CharSequence commaSeparate(CsvValueEscaper<T> escaper, T... values) {
+        StringBuilder sb = new StringBuilder(values.length * VALUE_LENGTH_ESTIMATE);
+        if (values.length > 0) {
+            int end = values.length - 1;
+            for (int i = 0; i < end; i++) {
+                sb.append(escaper.escape(values[i])).append(COMMA);
+            }
+            sb.append(escaper.escape(values[end]));
+        }
+        return sb;
+    }
+
+    private <T> CharSequence commaSeparate(CsvValueEscaper<T> escaper, Iterable<? extends T> values) {
+        StringBuilder sb = new StringBuilder();
```

Comment:
Can we check if values is instanceof Collection and if so use it to calculate proper size of StringBuilder.

Something like:

```java
final StringBuilder sb = values instanceof Collection ? new StringBuilder(((Collection) values).size() * VALUE_LENGTH_ESTIMATE) : new StringBuilder);
```